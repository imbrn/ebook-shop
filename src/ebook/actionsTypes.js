export const FETCH_ALL_EBOOKS = "FETCH_ALL_EBOOKS";
