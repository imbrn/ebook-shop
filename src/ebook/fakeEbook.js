import Ebook from "./Ebook";

export function fakeEbook(id) {
  return new Ebook({
    id,
    name: `Ebook ${id}`,
    description: `The ebook ${id}`,
    price: id * 10000
  });
}

export function fakeEbooks(amount = 10, firstId = 1) {
  return Array(amount)
    .fill()
    .map(() => fakeEbook(firstId++));
}
