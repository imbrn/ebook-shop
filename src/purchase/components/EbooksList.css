.ebooksList {
  li {
    margin: 40px 0;
    list-style: none;
  }
}

.ebookItem {
  display: flex;
  align-items: center;

  & > img {
    margin-left: 40px;
  }

  & > .ebookItemTitle {
    margin-left: 25px;
    display: flex;
    flex-direction: column;
    flex-grow: 1;

    & > * {
      max-width: 300px;
    }
  }
}

.totalBox {
  padding: 38px 0;
  display: flex;
  justify-content: space-between;
}
