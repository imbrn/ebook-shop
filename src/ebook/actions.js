import { getService } from "../service";
import { FETCH_ALL_EBOOKS } from "./actionsTypes";

export function fetchAllEbooks() {
  return dispatch => {
    dispatch(onFetchAllEbooks("request"));
    return getService()
      .fetchAllEbooks()
      .then(ebooks => dispatch(onFetchAllEbooks("success", { ebooks })));
  };
}

function onFetchAllEbooks(status, props) {
  return {
    type: FETCH_ALL_EBOOKS,
    status,
    ...props
  };
}
