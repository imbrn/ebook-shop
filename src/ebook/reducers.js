import { FETCH_ALL_EBOOKS } from "./actionsTypes";

export function ebooks(state = [], action) {
  switch (action.type) {
    case FETCH_ALL_EBOOKS:
      return action.ebooks;
    default:
      return state;
  }
}
