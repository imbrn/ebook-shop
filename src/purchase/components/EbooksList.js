import React from "react";
import { Text, CheckBox, Image, Separator } from "../../common/components";
import formatPrice from "../../common/utils/formatPrice";
import { connect } from "react-redux";
import { toggleEbookSelection } from "../actions";
import styles from "./EbooksList.css";

export const EbooksList = ({
  ebooks,
  purchase,
  formatPrice,
  toggleEbookSelection
}) => {
  return (
    <div className={styles.ebooksList}>
      <ul>
        {ebooks.items.map(ebook => (
          <li key={ebook.id}>
            <EbookItem
              ebook={ebook}
              selected={purchase.ebooks.includes(ebook)}
              formatPrice={formatPrice}
              onRequestSelectChange={() => toggleEbookSelection(ebook)}
            />
          </li>
        ))}
      </ul>

      <Separator />

      <div className={styles.totalBox}>
        <Text bold size="medium">
          TOTAL
        </Text>
        <Text bold size="medium">
          {formatPrice(totalSum(purchase.ebooks))}
        </Text>
      </div>
    </div>
  );
};

function totalSum(ebooks) {
  return ebooks.reduce((total, ebook) => {
    return total + ebook.price;
  }, 0);
}

EbooksList.defaultProps = {
  ebooks: {},
  purchase: {},
  formatPrice,
  toggleEbookSelection: () => {}
};

export const EbookItem = ({
  ebook,
  selected,
  formatPrice,
  onRequestSelectChange
}) => (
  <div className={styles.ebookItem}>
    <CheckBox
      value={ebook.id}
      checked={selected}
      onChange={e => onRequestSelectChange(!selected, e)}
    />
    <Image src={`public/images/ebooks/${ebook.id}.png`} />
    <div className={styles.ebookItemTitle}>
      <Text bold>{ebook.name}</Text>
      <Text>{ebook.description}</Text>
    </div>
    <Text bold size="medium">
      {formatPrice(ebook.price)}
    </Text>
  </div>
);

EbookItem.defaultProps = {
  formatPrice,
  onRequestSelectChange: () => {}
};

const mapStateToProps = state => ({
  ebooks: state.ebooks,
  purchase: state.purchase
});

const mapDispatchToProps = dispatch => ({
  toggleEbookSelection: ebook => dispatch(toggleEbookSelection(ebook))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EbooksList);
