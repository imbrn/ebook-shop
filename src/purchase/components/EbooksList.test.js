import React from "react";
import { render, cleanup, fireEvent } from "react-testing-library";
import { EbooksList, EbookItem } from "./EbooksList";
import Ebook from "../../ebook/Ebook";
import { fakeEbooks, fakeEbook } from "../../ebook/fakeEbook";

afterEach(cleanup);

describe("EbooksList", () => {
  test("renders correctly", () => {
    const ebooks = fakeEbooks(10);
    const selectedEbooks = [ebooks[0], ebooks[1], ebooks[5], ebooks[8]];
    const { container } = render(
      <EbooksList ebooks={ebooks} selectedEbooks={selectedEbooks} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("total price sum", () => {
    const ebooks = [
      new Ebook({ id: 1, name: "One", price: 1050 }),
      new Ebook({ id: 2, name: "Two", price: 3209 }),
      new Ebook({ id: 3, name: "Three", price: 1125 })
    ];
    const { getByText } = render(
      <EbooksList ebooks={ebooks} selectedEbooks={[ebooks[0], ebooks[2]]} />
    );
    expect(getByText("R$ 21,75")).toBeDefined();
  });

  test("handle ebook toggle selection", () => {
    const toggleEbookSelection = jest.fn();
    const ebooks = fakeEbooks(5);
    const { getByText } = render(
      <EbooksList ebooks={ebooks} toggleEbookSelection={toggleEbookSelection} />
    );
    const ebookEl = getByText(ebooks[1].name);
    const ebookCheckBox = ebookEl.parentElement.parentElement.querySelector(
      "[type='checkbox']"
    );
    fireEvent.click(ebookCheckBox);
    expect(toggleEbookSelection.mock.calls).toEqual([[ebooks[1]]]);
  });
});

describe("EbookItem", () => {
  test("renders correctly", () => {
    const ebook = fakeEbook(15);
    const { container } = render(<EbookItem ebook={ebook} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
