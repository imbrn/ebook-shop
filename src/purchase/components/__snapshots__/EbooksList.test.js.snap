// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EbookItem renders correctly 1`] = `
.c0 {
  position: relative;
  width: 20px;
  height: 20px;
  border: 2px solid black;
}

.c2 {
  position: absolute;
  opacity: 0;
  width: 100%;
  height: 100%;
  cursor: pointer;
}

.c1 {
  position: absolute;
  width: 100%;
  height: 100%;
}

.c3 {
  font-weight: bold;
  font-size: 14px;
}

.c4 {
  font-size: 14px;
}

.c5 {
  font-weight: bold;
  font-size: 18px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    />
    <input
      class="c2"
      type="checkbox"
      value="15"
    />
  </div>
  <img
    class=""
    src="public/images/ebooks/15.png"
  />
  <div>
    <span
      class="c3"
    >
      Ebook 15
    </span>
    <span
      class="c4"
    >
      The ebook 15
    </span>
  </div>
  <span
    class="c5"
  >
    R$ 1500,00
  </span>
</div>
`;

exports[`EbooksList renders correctly 1`] = `
.c0 {
  position: relative;
  width: 20px;
  height: 20px;
  border: 2px solid black;
}

.c2 {
  position: absolute;
  opacity: 0;
  width: 100%;
  height: 100%;
  cursor: pointer;
}

.c1 {
  position: absolute;
  width: 100%;
  height: 100%;
}

.c6 {
  position: absolute;
  width: 100%;
  height: 100%;
  background: black;
  width: 2px;
  height: 2px;
  left: 3px;
  top: 7px;
  box-shadow: 2px 0 0 black,4px 0 0 black,4px -2px 0 black, 4px -4px 0 black,4px -6px 0 black,4px -8px 0 black;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

.c3 {
  font-weight: bold;
  font-size: 14px;
}

.c4 {
  font-size: 14px;
}

.c5 {
  font-weight: bold;
  font-size: 18px;
}

.c7 {
  border: none;
  height: 2px;
  background: #dedede;
}

<div>
  <ul>
    <li>
      <div>
        <div
          class="c0"
        >
          <div
            class="c1"
          />
          <input
            class="c2"
            type="checkbox"
            value="1"
          />
        </div>
        <img
          class=""
          src="public/images/ebooks/1.png"
        />
        <div>
          <span
            class="c3"
          >
            Ebook 1
          </span>
          <span
            class="c4"
          >
            The ebook 1
          </span>
        </div>
        <span
          class="c5"
        >
          R$ 100,00
        </span>
      </div>
    </li>
    <li>
      <div>
        <div
          class="c0"
        >
          <div
            class="c6"
          />
          <input
            checked=""
            class="c2"
            type="checkbox"
            value="2"
          />
        </div>
        <img
          class=""
          src="public/images/ebooks/2.png"
        />
        <div>
          <span
            class="c3"
          >
            Ebook 2
          </span>
          <span
            class="c4"
          >
            The ebook 2
          </span>
        </div>
        <span
          class="c5"
        >
          R$ 200,00
        </span>
      </div>
    </li>
    <li>
      <div>
        <div
          class="c0"
        >
          <div
            class="c1"
          />
          <input
            class="c2"
            type="checkbox"
            value="3"
          />
        </div>
        <img
          class=""
          src="public/images/ebooks/3.png"
        />
        <div>
          <span
            class="c3"
          >
            Ebook 3
          </span>
          <span
            class="c4"
          >
            The ebook 3
          </span>
        </div>
        <span
          class="c5"
        >
          R$ 300,00
        </span>
      </div>
    </li>
    <li>
      <div>
        <div
          class="c0"
        >
          <div
            class="c6"
          />
          <input
            checked=""
            class="c2"
            type="checkbox"
            value="4"
          />
        </div>
        <img
          class=""
          src="public/images/ebooks/4.png"
        />
        <div>
          <span
            class="c3"
          >
            Ebook 4
          </span>
          <span
            class="c4"
          >
            The ebook 4
          </span>
        </div>
        <span
          class="c5"
        >
          R$ 400,00
        </span>
      </div>
    </li>
  </ul>
  <hr
    class="c7"
  />
  <div>
    <span
      class="c5"
    >
      TOTAL
    </span>
    <span
      class="c5"
    >
      R$ 600,00
    </span>
  </div>
</div>
`;
